### PR TITLE
Approved: Add active effects

### PR DESCRIPTION
**Pull request by JFranklin3000**
Active Effects are active on character actors and on perks and hang-ups for now. 

test cases on both user and item
add effect that increases Essences (system.essences. Add X)
add effect that increases Defenses in same way (system.bonuses. Add X)
add effect that sets skill to value (system.skills...shift set d8 or whatever)
add effect that sets bonus to skill (system.skills...modifier Add X)

### CODE REVIEW

File: templates/actor/parts/actor-active-effects.hbs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+                    <i class="fas fa-trash"></i>
+                </a>
+                {{/if}}
+            </div>
+        </li>
+    {{/each}}
+    </ol>
+{{/unless}}
+{{/each}}
+</ol>
```
Comment (phildominguez) [inline templates/actor/parts/actor-active-effects.hbs:53]: Is it possible to style the containers to be like the collapsible item containers we have?

File: module/sheets/item-sheet.mjs [hunk 2]
```diff
@@ -44,10 +45,13 @@ export class Essence20ItemSheet extends ItemSheet {
     if (actor) {
       context.rollData = actor.getRollData();
     }
+    // Prepare active effects
+    context.effects = prepareActiveEffectCategories(this.object.effects);
```
Comment (phildominguez) [inline module/sheets/item-sheet.mjs:49]: I get this when I open the sheet and whatnot:
image
I think just make this `e.disabled` here https://github.com/WookieeMatt/Essence20/blob/main/module/helpers/effects.mjs#L58

**Review by phildominguez — CHANGES_REQUESTED** [review]
Haven't fully tested yet, but I know you like syntax feedback!

**Comment by JFranklin3000** [comment]
test cases on both user and item
add effect that increases Essences (system.essences.<name> Add X)
add effect that increases Defenses in same way (system.bonuses.<name> Add X)
add effect that sets skill to value (system.skills.<essence>.<skill>.shift set d8 or whatever)
add effect that sets bonus to skill (system.skills.<essence>.<skill>.modifier Add X)

File: lang/en.json [hunk 3]
```diff
@@ -96,6 +96,13 @@
     "DefenseUnmorphed": "Unmorphed",
     "DefenseValues": "Values",
     "DefenseWillpower": "Willpower",
+    "Duration": "Duration",
+    "EffectAdd": "Add",
+    "EffectCreate": "Create Effect",
+    "EffectDelete": "Delete Effect",
+    "EffectDisable": "Disable Effect",
+    "EffectEdit": "Edit Effect",
+    "EffectEnable": "Enable Effect",
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply lang/en.json:105]: renamed to ActiveEffectAdd for example. Duration just got missed.

File: module/sheets/item-sheet.mjs [hunk 4]
```diff
@@ -44,10 +45,13 @@ export class Essence20ItemSheet extends ItemSheet {
...
 
     // Add the actor's data to context.data for easier access, as well as flags.
     context.system = itemData.system;
     context.flags = itemData.flags;
+
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply module/sheets/item-sheet.mjs:54]: removed

File: module/sheets/item-sheet.mjs [hunk 5]
```diff
@@ -2,6 +2,7 @@
  * Extend the basic ItemSheet with some very simple modifications
  * @extends {ItemSheet}
  */
+import { onManageActiveEffect, prepareActiveEffectCategories } from "../helpers/effects.mjs";
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply module/sheets/item-sheet.mjs:5]: Moved.

File: module/sheets/item-sheet.mjs [hunk 6]
```diff
@@ -44,10 +45,13 @@ export class Essence20ItemSheet extends ItemSheet {
     if (actor) {
       context.rollData = actor.getRollData();
     }
+    // Prepare active effects
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply module/sheets/item-sheet.mjs:48]: Added.

File: templates/item/parts/item-active-effects.hbs [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+                    <i class="fas fa-trash"></i>
+                </a>
+                {{/if}}
+            </div>
+        </li>
+    {{/each}}
+    </ol>
+{{/unless}}
+{{/each}}
+</ol>
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply templates/item/parts/item-active-effects.hbs:53]: Fixed indenting

File: templates/item/parts/item-active-effects.hbs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+            <div class="item-name effect-name flexrow">
+                <img class="item-image" src="{{effect.icon}}"/>
+                <h4>{{effect.label}}</h4>
+            </div>
+            <div class="effect-source">{{effect.sourceName}}</div>
+            <div class="effect-duration">{{effect.duration.label}}</div>
+            <div class="item-controls effect-controls flexrow">
+                {{#if @root.editable}}
+                <a class="effect-control" data-action="toggle"
+                   data-tooltip="{{#if effect.disabled}}E20.EffectEnable{{else}}E20.EffectDisable{{/if}}">
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply templates/item/parts/item-active-effects.hbs:37]: yep, not taking credit for it I borrowed that from the 5e code.

File: templates/actor/parts/actor-active-effects.hbs (hunk 1 above)
Comment (JFranklin3000) [reply templates/actor/parts/actor-active-effects.hbs:53]: I would think so I asked @guswadsworth12 to take a look as well.

File: templates/actor/actor-powerRanger-sheet.hbs [hunk 9]
```diff
@@ -54,6 +55,11 @@
       {{> "systems/essence20/templates/actor/parts/actor-background.hbs"}}
     </div>
 
+    {{!-- Effects Section --}}
+    <div class="tab background flexcol" data-group="primary" data-tab="effects"></div>
+      {{> "systems/essence20/templates/actor/parts/actor-active-effects.hbs"}}
+    </div>
+
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply templates/actor/actor-powerRanger-sheet.hbs:62]: how do you mean that is the div for the tab, not sure why the list gets created outside of the tab div though.

File: templates/actor/actor-giJoe-sheet.hbs [hunk 10]
```diff
@@ -47,6 +48,11 @@
     <div class="tab background flexcol" data-group="primary" data-tab="background">
       {{> "systems/essence20/templates/actor/parts/actor-background.hbs"}}
     </div>
+    
+    {{!-- Effects Tab --}}
+    <div class="tab background flexcol" data-group="primary" data-tab="effects"></div>
```
Comment (phildominguez) [inline templates/actor/actor-giJoe-sheet.hbs:53]: `background` here should be `effects` to match `data-tab` above. Needs to be fixed for all sheets.

There's also a trailing `</div>` on this line that's causing it to show up on every tab.

File: templates/actor/parts/actor-active-effects.hbs (hunk 1 above)
Comment (phildominguez) [reply templates/actor/parts/actor-active-effects.hbs:53]: We spoke offline but just to note, I only mean the rounded border style and whatnot, not actually make it work with collapsible-item-container just yet

File: templates/item/item-hangUp-sheet.hbs [hunk 11]
```diff
@@ -5,6 +5,10 @@
 
   {{!-- Item Sheet Body --}}
   <section class="sheet-body item-sheet">
+        
+    {{!-- Effects Section --}}
+    {{> "systems/essence20/templates/item/parts/item-active-effects.hbs"}}
+
```
Comment (phildominguez) [inline templates/item/item-hangUp-sheet.hbs:11]: You're probably aware:
image
Should we go with an Effects tab like the actor sheets?

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: templates/actor/parts/actor-active-effects.hbs (hunk 1 above)
Comment (JFranklin3000) [reply templates/actor/parts/actor-active-effects.hbs:53]: Border added

File: module/sheets/item-sheet.mjs [hunk 12]
```diff
@@ -1,3 +1,4 @@
+import { onManageActiveEffect, prepareActiveEffectCategories } from "../helpers/effects.mjs";
```
[the remark this answers is not in the document]
Comment (JFranklin3000) [reply module/sheets/item-sheet.mjs:1]: Added

File: templates/actor/parts/actor-active-effects.hbs (hunk 1 above)
Comment (phildominguez) [reply templates/actor/parts/actor-active-effects.hbs]: image
Sorry for the confusion, this is what I meant. Remove the outer border, round the ones around the headers, and be sure they use the actor's color. I think make the font larger so it matches what we use elsewhere as well.

File: templates/item/item-perk-sheet.hbs [hunk 13]
```diff
@@ -11,9 +11,12 @@
...
-
+    
+    {{!-- Effects Section --}}
+    {{> "systems/essence20/templates/item/parts/item-active-effects.hbs"}}
+    
     {{!-- Description Section --}}
     {{> "systems/essence20/templates/item/parts/item-description.hbs"}}
   </section>
-
+
```
Comment (phildominguez) [inline templates/item/item-perk-sheet.hbs:21]: Remove space 🙂

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: sass/assets/_headers.scss [hunk 14]
```diff
@@ -60,6 +60,8 @@
   background: v.$background-n;
   border: v.$border;
   font-weight: bold;
+  border-radius: v.$border-radius;
+  box-shadow: v.$boxshadow;
```
Comment (phildominguez) [inline sass/assets/_headers.scss:64]: I think you didn't run `sass sass/essence20.scss css/essence20.css`, so this was never applied to the CSS.

File: templates/item/item-altmode-sheet.hbs [hunk 15]
```diff
@@ -3,59 +3,76 @@
...
+      {{#*inline "item-field-inputs"}}
+      <input type="number" name="system.altModeFirepoints" value="{{system.altModeFirepoints}}" />
+      {{/inline}}
+      {{/"systems/essence20/templates/item/parts/item-sheet-field.hbs"}}
+    </div>
+
+    {{!-- Effects Tab --}}
+    <div class="tab skills flexcol" data-group="primary" data-tab="effects">
+      {{> "systems/essence20/templates/item/parts/item-active-effects.hbs"}}
+    </div>
```
Comment (phildominguez) [inline templates/item/item-altmode-sheet.hbs:74]: image
With the default item sheet width, this does not look great

File: templates/actor/parts/actor-active-effects.hbs [hunk 16]
```diff
@@ -0,0 +1,55 @@
+<div class="stats-container defense-container" style="border-color: {{system.color}};">
```
Comment (phildominguez) [inline templates/actor/parts/actor-active-effects.hbs:1]: Don't think we need this border

File: templates/actor/parts/actor-active-effects.hbs [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+  <ol class="items-list effects-list">
+  {{#each effects as |section sid|}}
+  {{#unless section.hidden}}
+    <li class="items-header flexrow" data-effect-type="{{section.type}}">
```
Comment (phildominguez) [inline templates/actor/parts/actor-active-effects.hbs:5]: Make the border color the actor's color

File: templates/item/item-altmode-sheet.hbs [hunk 18]
```diff
@@ -3,59 +3,76 @@
   {{!-- Item Sheet Header --}}
   {{> "systems/essence20/templates/item/parts/item-header.hbs"}}
 
+  {{!-- Tabs --}}
+  <nav class="sheet-tabs tabs" data-group="primary">
+    <a class="item" style="border-color: {{system.color}};" data-tab="description"> {{ localize "E20.TabDescription" }}</a>
+    <a class="item" style="border-color: {{system.color}};" data-tab="details">{{ localize "E20.TabDetails" }}</a>
```
Comment (phildominguez) [inline templates/item/item-altmode-sheet.hbs:9]: `E20.*` strings aren't defined

I also would not split up the main tab into description and details in this PR. It's not really related to this work and is something we should discuss first. If we want to go for it, we can do it for all items at once.

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: templates/item/item-altmode-sheet.hbs [hunk 19]
```diff
@@ -3,59 +3,74 @@
   {{!-- Item Sheet Header --}}
   {{> "systems/essence20/templates/item/parts/item-header.hbs"}}
 
+  {{!-- Tabs --}}
+  <nav class="sheet-tabs tabs" data-group="primary">
+    <a class="item" style="border-color: {{system.color}};" data-tab="details">{{ localize "E20.TabMain" }}</a>
+    <a class="item" style="border-color: {{system.color}};" data-tab="effects">{{ localize "E20.TabEffects" }}</a>
```
Comment (phildominguez) [inline templates/item/item-altmode-sheet.hbs:9]: `border-color` not needed here

File: templates/item/parts/item-active-effects.hbs [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+
+  <ol class="item-list">
+  {{#each section.effects as |effect|}}
+    <li class="item effect flexrow" data-effect-id="{{effect.id}}">
+      <div class="item-name effect-name flexrow">
+        <img class="item-image" src="{{effect.icon}}"/>
+        <h4>{{effect.label}}</h4>
+      </div>
+      <div class="effect-source">{{effect.sourceName}}</div>
+      <div class="effect-duration">{{effect.duration.label}}</div>
```
Comment (phildominguez) [inline templates/item/parts/item-active-effects.hbs:32]: `effect-source` and `effect-duration` don't seem to be needed. Let's remove them in the hbs and scss

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: templates/item/parts/item-active-effects.hbs (hunk 20 above)
Comment (JFranklin3000) [reply templates/item/parts/item-active-effects.hbs:32]: I didn't add them to the scss they were already there. but I have no issue removing if that won't affect anything else.

File: templates/item/item-altmode-sheet.hbs (hunk 19 above)
Comment (JFranklin3000) [reply templates/item/item-altmode-sheet.hbs:9]: Removed.

File: templates/actor/parts/actor-active-effects.hbs (hunk 17 above)
Comment (JFranklin3000) [reply templates/actor/parts/actor-active-effects.hbs:5]: Added on the Item Sheet.

File: templates/item/parts/item-active-effects.hbs (hunk 20 above)
Comment (phildominguez) [reply templates/item/parts/item-active-effects.hbs:32]: Ohh okay. Yeah I feel like if it doesn't do anything let's remove it from the hbs. I'm fine with leaving it in the scss.

**Comment by guswadsworth12** [comment]
Bug Identified:

Can not change the number of turns that an effect lasts for. Rounds works though. 
image

File: sass/items/_item-styling.scss [hunk 21]
```diff
@@ -108,12 +110,14 @@
   height: 28px;
   margin: 2px 0;
   padding: 0;
+  padding-bottom: 5px;
```
Comment (phildominguez) [inline sass/items/_item-styling.scss:113]: image
I think we want to remove this bottom padding

**Review by phildominguez — CHANGES_REQUESTED** [review]

File: sass/items/_item-styling.scss (hunk 21 above)
Comment (JFranklin3000) [reply sass/items/_item-styling.scss]: this has been removed.

**Review by phildominguez — APPROVED** [review]

**Review by guswadsworth12 — APPROVED** [review]
Looks and works good approved.

**Pull request merged**